File: 1995-finding-pairs-with-a-certain-sum/finding-pairs-with-a-certain-sum.py
class FindSumPairs:
    def __init__(self, nums1: List[int], nums2: List[int]):
        self.nums1 = nums1
        self.nums2 = nums2

        self.cache = defaultdict(int)
        for num in nums2:
            self.cache[num] += 1

    def add(self, index: int, val: int) -> None:
        prev = self.nums2[index]
        _next = prev + val
        self.cache[prev] -= 1
        self.nums2[index] = _next
        self.cache[_next] += 1

    def count(self, tot: int) -> int:
        count = 0
        for num in self.nums1:
            diff = tot - num
            count += self.cache[diff]
        return count



# Your FindSumPairs object will be instantiated and called as such:
# obj = FindSumPairs(nums1, nums2)
# obj.add(index,val)
# param_2 = obj.count(tot)

File: 1995-finding-pairs-with-a-certain-sum/finding-pairs-with-a-certain-sum.ts
class FindSumPairs {
    #nums1: number[]
    #nums2: number[];
    #cache: Map<number, number>;
    constructor(nums1: number[], nums2: number[]) {
        this.#nums1 = nums1;
        this.#nums2 = nums2;

        this.#cache = new Map();

        for (const num of nums2) {
            const count = this.#cache.get(num) || 0;
            this.#cache.set(num, count + 1);
        }
    }

    add(index: number, val: number): void {
        const prev = this.#nums2[index];
        const next = prev + val;

        this.#nums2[index] += val;
        this.#cache.set(prev, this.#cache.get(prev) - 1);
        this.#cache.set(next, (this.#cache.get(next) || 0) + 1);
    }

    count(tot: number): number {
        let cnt = 0;
        for (const num of this.#nums1) {
            const diff = tot - num;

            cnt += (this.#cache.get(diff) || 0);
        } 
        return cnt;
    }
}

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * var obj = new FindSumPairs(nums1, nums2)
 * obj.add(index,val)
 * var param_2 = obj.count(tot)
 */